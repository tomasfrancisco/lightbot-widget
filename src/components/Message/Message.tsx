import * as React from "react";
import styled, { ThemedStyledProps, withTheme } from "styled-components";

const getBackgroundColor = (props: ThemedStyledProps<MessageTagProps, any>) =>
  props.isBotMessage
    ? props.theme.botSaysBackgroundColor || "#F2F2F2"
    : props.theme.userSaysBackgroundColor || "#4facfe";

const getTextColor = (props: ThemedStyledProps<MessageTagProps, any>) =>
  props.isBotMessage
    ? props.theme.botSaysTextColor || "#4C4C4C"
    : props.theme.userSaysTextColor || "white";

export type MessageTagProps = {
  isBotMessage?: boolean;
};

const MessageTag = withTheme(styled<MessageTagProps, "div">("div")`
  margin: 1px 0;
  padding: 10px 15px;
  background-color: ${props => getBackgroundColor(props)};
  border-radius: 25px;
  color: ${props => getTextColor(props)};
  font-family: "Assistant", sans-serif;
  max-width: 290px;
  float: ${props => (props.isBotMessage ? "left" : "right")};
`);

const MessageWrapper = styled("div")`
  display: block;
`;

export interface MessageProps {
  children?: any;
  isBotMessage?: boolean;
}

export const Message = (props: MessageProps) => {
  const { children, isBotMessage } = props;
  return (
    <MessageWrapper className="message-wrapper">
      <MessageTag className="message-tag" isBotMessage={isBotMessage}>
        {children}
      </MessageTag>
    </MessageWrapper>
  );
};
