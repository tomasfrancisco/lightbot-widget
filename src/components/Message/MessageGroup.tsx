import * as React from "react";
import styled from "styled-components";

import { MessageProps } from "./Message";

type MessageGroupWrapperProps = {
  isRtl?: boolean;
};

export const MessageGroupWrapper = styled<MessageGroupWrapperProps, "div">("div")`
  display: flex;
  flex-grow: 1;
  flex-shrink: 0;
  flex-direction: row;
  justify-content: ${props => (props.isRtl ? "flex-end" : "flex-start")};

  &:not(:first-of-type) {
    margin-top: 10px;
  }

  & > .avatar-container {
    order: ${props => (props.isRtl ? 2 : 1)};
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    ${props => (props.isRtl ? "margin-left" : "margin-right")}: 5px;

    & > .avatar {
      display: flex;
      justify-content: center;
      align-items: center;
      width: 30px;
      height: 30px;
      border-radius: 25px;
      margin-bottom: 5px;
      background-color: tomato;
      color: white;
    }
  }

  & > .message-container {
    order: ${props => (props.isRtl ? 1 : 2)};
    display: flex;
    flex-flow: column;
  }


  & .message-wrapper {
    &:not(:only-of-type) {
      &:first-of-type .message-tag {
        border-bottom-${props => (props.isRtl ? "right" : "left")}-radius: 4px;
      }

      &:not(:first-of-type):not(:last-of-type) .message-tag {
        border-bottom-${props => (props.isRtl ? "right" : "left")}-radius: 4px;
        border-top-${props => (props.isRtl ? "right" : "left")}-radius: 4px;
      }

      &:last-of-type .message-tag {
        border-top-${props => (props.isRtl ? "right" : "left")}-radius: 4px;
      }
    }
  }
`;

interface MessageGroupProps extends MessageProps {
  children?: any;
  isRtl?: boolean;
  showAvatar?: boolean;
  avatarName?: string;
  onClick?: (event) => void;
}

const passPropsDown = ({ children, ...props }: MessageGroupProps) =>
  React.Children.map(children, child => React.cloneElement(child as any, props));

const renderAvatar = avatarName => {
  return (
    <div className="avatar-container">
      <div className="avatar">{avatarName && avatarName.charAt(0)}</div>
    </div>
  );
};

export const MessageGroup = (props: MessageGroupProps) => {
  const { isRtl, showAvatar, avatarName, onClick, ...rest } = props;
  return (
    <MessageGroupWrapper isRtl={isRtl} onClick={onClick}>
      {showAvatar && renderAvatar(avatarName)}
      <div className="message-container">{passPropsDown(rest)}</div>
    </MessageGroupWrapper>
  );
};
