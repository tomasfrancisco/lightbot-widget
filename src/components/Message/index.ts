export * from "./Message";
export * from "./MessageGroup";
