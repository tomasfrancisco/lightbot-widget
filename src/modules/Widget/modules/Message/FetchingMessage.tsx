import styled from "styled-components";

export const FetchingMessage = styled("div")`
	display: inline-block;
  position: relative;
  width: 64px;
  height: 64px;

	&:before,
	&:after {
		content: "";
		position: absolute;
    border: 4px solid #fff;
    opacity: 1;
    border-radius: 50%;
    animation: lds-ripple 1s cubic-bezier(0, 0.2, 0.8, 1) infinite;
	}

	&:after {
		animation-delay: -0.5s;
	}
}

@keyframes lds-ripple {
  0% {
    top: 28px;
    left: 28px;
    width: 0;
    height: 0;
    opacity: 1;
  }
  100% {
    top: -1px;
    left: -1px;
    width: 58px;
    height: 58px;
    opacity: 0;
  }
`;
