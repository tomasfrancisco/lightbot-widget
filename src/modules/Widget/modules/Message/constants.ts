import { MessageSenderEnum, MessageTypeEnum } from "./Message.type";

export const fetchingMessage = {
  label: "Typing...",
  sender: MessageSenderEnum.BOT,
  type: MessageTypeEnum.FETCHING,
};
