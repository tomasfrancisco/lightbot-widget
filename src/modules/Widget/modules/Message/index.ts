export * from "./Message.module";
export * from "./Message.type";
