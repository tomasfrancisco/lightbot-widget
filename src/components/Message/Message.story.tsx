import { storiesOf } from "@storybook/react";
import * as React from "react";
import { ThemeProviderDecorator } from "../../modules/ThemeProvider/ThemeProvider";
import { Message } from "./Message";
import { MessageGroup } from "./MessageGroup";

storiesOf("Message", module)
  .addDecorator(ThemeProviderDecorator)
  .add("default", () => <Message>Default Message</Message>)
  .add("message group", () => (
    <MessageGroup showAvatar={true}>
      <Message>Example #1</Message>
      <Message>Example #2</Message>
      <Message>Example #3</Message>
    </MessageGroup>
  ))
  .add("message group RTL", () => (
    <MessageGroup isRtl={true}>
      <Message>Example #1</Message>
      <Message>Example #2</Message>
      <Message>Example #3</Message>
    </MessageGroup>
  ));
