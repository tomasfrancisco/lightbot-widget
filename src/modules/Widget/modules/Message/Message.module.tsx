import { Button } from "components/Button";
import { Link } from "components/Link";
import { Message } from "components/Message";
import * as React from "react";
import Linkify from "react-linkify";
import { v4 as uuid } from "uuid";

import { FetchingMessage } from "./FetchingMessage";
import {
   JumpMessage, LinkMessage, MessageSenderEnum, MessageType, MessageTypeEnum,
} from "./Message.type";

export class MessageModule extends React.Component<{
  message: MessageType;
  sender: MessageSenderEnum;
  touched?: boolean;
  onClick?: (message: string, label?: string, type?: MessageTypeEnum) => void;
}> {
  public render() {
    const { sender, ...props } = this.props;
    return (
      <Message isBotMessage={sender === MessageSenderEnum.BOT} {...props}>
        {this.decodeMessage(this.props.message)}
      </Message>
    );
  }

  private decodeMessage = (message: MessageType) => {
    const { touched } = this.props;

    switch (message.type) {
      case MessageTypeEnum.PLAIN:
        return (
          <span key={uuid()}>
            <Linkify properties={{ target: "_blank" }}>{message.label}</Linkify>
          </span>
        );
      case MessageTypeEnum.JUMP:
        const jumpMessage = message as JumpMessage;
        return jumpMessage.jumps.map(jump => (
          <Button
            key={uuid()}
            negative={true}
            onClick={this.getOnButtonClick(jump.event, jump.label)}
            touched={touched ? touched : false}
          >
            {jump.label}
          </Button>
        ));
      case MessageTypeEnum.LINK:
        const linkMessage = message as LinkMessage;
        return (
          <Link key={uuid()} href={linkMessage.link} target="_blank">
            {linkMessage.label || linkMessage.link}
          </Link>
        );
      case MessageTypeEnum.DECORATED:
        return <p key={uuid()}>{this.decodeDecoratedType(message)}</p>;
      case MessageTypeEnum.FETCHING:
        return <FetchingMessage />;
      default:
        return null;
    }
  };

  private getOnButtonClick = (event: string, label?: string) => {
    const { onClick } = this.props;

    if (onClick) {
      return () => onClick(event, label, MessageTypeEnum.JUMP);
    }

    return undefined;
  };

  private decodeDecoratedType = message => {
    const messageSubstrings = message.label.split(/{\d}/);
    return messageSubstrings.reduce(
      (result, substring, i) => [
        ...result,
        this.decodeMessage({
          label: substring,
          sender: message.sender,
          type: MessageTypeEnum.PLAIN,
        }),
        i < messageSubstrings.length - 1 ? this.decodeMessage(message.objects[i]) : null,
      ],
      [],
    );
  };
}
