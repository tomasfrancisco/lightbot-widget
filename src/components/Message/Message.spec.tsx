import { Message } from "components/Message/Message";
import { shallow } from "enzyme";
import * as React from "react";

describe("modules/Widget/Window", () => {
  let component;

  beforeEach(() => {
    component = shallow(<Message />);
  });

  it("renders", () => {
    expect(component).toMatchSnapshot();
  });
});
