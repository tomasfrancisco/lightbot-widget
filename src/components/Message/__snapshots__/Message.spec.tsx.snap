// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/Widget/Window renders 1`] = `
<styled.div
  className="message-wrapper"
>
  <WithTheme(styled.div)
    className="message-tag"
  />
</styled.div>
`;
